src/editor.js

```diff
@@ -1257,6 +1296,124 @@ export class OdooEditor extends EventTarget {
         }
     }
 
+    // PASTING / DROPPING
+
+    /**
+     * Prepare clipboard data (text/html) for safe pasting into the editor.
+     *
+     * @private
+     * @param {string} clipboardData
+     * @returns {string}
+     */
+     _prepareClipboardData(clipboardData) {
+        const container = document.createElement('fake-container');
+        container.innerHTML = clipboardData;
+        return this._cleanForPaste(container).innerHTML;
+    }
+    /**
+     * Prepare clipboard data (text/plain) for safe pasting into the editor.
+     *
+     * @private
+     * @param {string} clipboardData
+     * @returns {string}
+     */
+    _prepareTextClipboardData(clipboardData) {
+        const isXML = !!clipboardData.match(/<[a-z]+[a-z0-9-]*( [^>]*)*>[\s\S\n\r]*<\/[a-z]+[a-z0-9-]*>/i);
+        const isJS = !isXML && !!clipboardData.match(/\(\);|this\.|self\.|function\s?\(|super\.|[a-z0-9]\.[a-z].*;/i);
+
+        const container = document.createElement('fake-container');
+        const pre = document.createElement('pre');
+        pre.innerHTML = clipboardData.trim()
+            .replace(/</g, '&lt;').replace(/>/g, '&gt;')
+            // Get that text as an array of text nodes separated by <br> where
+            // needed.
+            .replace(/(\n+)/g, '<br>');
+
+        if (isJS || isXML) {
+            container.appendChild(pre);
+        } else {
+            for (const node of pre.childNodes) {
+                container.appendChild(node);
+            };
+        }
+        return container.innerHTML;
+    }
+    /**
+     * Clean a container element's children (or the container itself if
+     * `justContainer` is true) for safely pasting. Cleaning an element involves
+     * unwrapping its contents if it's an illegal (blacklisted or not
+     * whitelisted) element, or removing its illegal attributes and classes.
+     * Return the cleaned element.
+     *
+     * @param {Node} container
+     * @param {boolean} [justContainer=false]
+     * @returns {Node}
+     */
+    _cleanForPaste(container, justContainer = false) {
```

Comment:
Not a fan of `justContainer` but can't think of anything better.

I could maybe pass an array of nodes to `_cleanForPaste` but I'm afraid that would look more complicated. Not sure though.

Reply:
I don't understand why this `justContainer` parameter is necessary. Is the function recursive if and only if this parameter is `true` ? If yes, then call it `recurse` or something like that. If not, then why not simply loop over the elements children at call site when you would call the function with `false` and call it for each child instead ?

Reply:
`justContainer === false` -> clean the container's children (not the container itself).
`justContainer === true` -> clean the container itself (not its children).

I also don't like it. I could instead systematically pass an array of nodes to clean (`justContainer ? [container] : [...container.childNodes]` basically).

Reply:
I would have moved the loop outside of `_cleanForPaste` and rename `containers` to `node`. This way, the method always cleans the given node and that's it. If you need to clear more nodes, call it on each node. It seems more elegant to me. What do you think ?